Lex/graph data by date range

LPeterson18 opened the pull request:
1. Add datePicker, spinner, timezone, and router dependencies
2. Allow frontend routing
3. Add Spinner, timepicker, inputs, and buttons components
4. Graph data based on start and end dates from inputs on sidebar.  Will also graph based on url param start and end dates.

algtan commented on src/client/components/Buttons.tsx:
```diff
@@ -0,0 +1,31 @@
+
+export default function MainButton(props: buttonProps): JSX.Element {
+  return (
+    <button style={mainButton} onClick={() => props.onClick()}>
```
Check if this component still works  if you change onClick from `onClick={() => props.onClick()}` to `onClick={props.onClick}`.

algtan commented on src/client/components/Sidebar.tsx:
```diff
@@ -1,7 +1,117 @@
+  dateStr: string,
+  timezoneOffset: number
+): Date {
+  // Find offset between local timezone and UTC in milliseconds
+  const localTimezoneOffset = new Date().getTimezoneOffset() * 60 * 1000
+  // Find offset between selected timezone and UTC in milliseconds
+  const selectedTimezoneOffset = timezoneOffset * 60 * 60 * 1000
+  // Return adjusted date from local timezone to selected timezone
+  return new Date(
+    new Date(dateStr).getTime() - localTimezoneOffset - selectedTimezoneOffset
+  )
```
See if you can change this from using JavaScript's Date object, to using date-fns instead. The `getTimezoneOffset()` method can return a number that is +/- 60 minutes because of Daylight Savings.

algtan commented on src/client/components/Sidebar.tsx:
```diff
@@ -1,7 +1,117 @@
+        <hr style={divider} />
+        <TimePicker
+          label="Start"
+          timePicker
```
Is `timePicker` undefined here? In the child component, `timePicker` is destructured to a default value of `false`.

algtan replied:
This is a boolean prop, which means `timePicker` is equal to true when it is added to the `TimePicker` component.

algtan commented on src/client/components/Timepicker.tsx:
```diff
@@ -0,0 +1,49 @@
+
+class TimePicker extends PureComponent<TimePickerProps, {}> {
+  render(): JSX.Element {
+    const { label, timePicker = false } = this.props
```
See comment on parent component regarding the `timePicker` prop.

algtan commented on src/client/components/GraphScene.tsx:
```diff
@@ -126,21 +131,54 @@ export const dataSelecter = (rawData: any, provider: string): any => {
+  const location = useLocation<Location>()
+
+  const getData = async (start: string, end: string): Promise<number> => {
+    let data
+    console.time('getData')
+    props.setLoading(true)
+    let response
+    const uri = `${window.location.origin}/v1/getDataAllPartners/?firstDate=${start}&secondDate=${end}`
```
All of this can probably be moved into the `try` block. Then instead of `let data` and `let response` outside of it, you can do `const response = await fetch(uri)` and `const data = await response.json()`.

algtan commented on src/client/components/GraphScene.tsx:
```diff
@@ -126,21 +131,54 @@ export const dataSelecter = (rawData: any, provider: string): any => {
+    try {
+      response = await fetch(uri)
+      data = await response.json()
+      props.setLoading(false)
+      setGraphData(data)
+    } catch (e) {
+      props.setLoading(false)
```
Instead of having `props.setLoading(false)` in both the try and catch blocks, try adding it once to a `finally` block. The `finally` will run whether whether an exception is thrown or caught.

algtan reviewed (changes requested):
Please see comments for feedback and requested changes.

LPeterson18 commented:
Made changes according to comments.

algtan commented on src/client/components/GraphScene.tsx:
```diff
@@ -136,17 +136,16 @@ export default function GraphScene(props: GraphSceneProps): JSX.Element {
-      props.setLoading(false)
+      const data = await response.json()
       setGraphData(data)
     } catch (e) {
```
Suggest adding a console.log so that you know an error was caught.

algtan reviewed (changes requested):
Looks good, I only have one comment.

LPeterson18 commented:
Added log to catch according to comments.

LPeterson18 commented on src/client/components/Buttons.tsx:
```diff
@@ -0,0 +1,31 @@
+  cursor: 'pointer' as 'pointer'
+}
+
+const mainButton = {
+  ...button,
+  overflow: 'hidden' as 'hidden',
+  marginTop: '12px',
+  marginLeft: '68px',
+  marginBottom: '12px',
+  color: 'white',
+  border: '1px solid white'
+}
```
Use theme file for px and other styling

LPeterson18 commented:
1. Converted fetch data function into hook separate from GraphScene component and restructured GraphScene
2. Removed Excess functionality from inputs
3. Converted inputs, spinner, and timepicker to hooks
4. Created clientUtils folder and file and moved various standalone functions there
5. Used themes and strings files for all styling constants

algtan commented on src/client/hooks/useFetchGraphData.ts:
```diff
@@ -0,0 +1,31 @@
+    end: string,
+    setLoading: React.Dispatch<any>
+  ): Promise<any> => {
+    const uri = `${window.location.origin}/v1/getDataAllPartners/?firstDate=${start}&secondDate=${end}`
```
I believe this line can be moved inside the `try`.

algtan commented on src/client/components/GraphScene.tsx:
```diff
@@ -133,39 +129,24 @@ export const dataSelecter = (rawData: any, provider: string): any => {
 export default function GraphScene(props: GraphSceneProps): JSX.Element {
   const [graphData, setGraphData] = useState([] as any)
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const location = useLocation<Location>()
```
May want to double-check this. Is `location` not used? If not, is line 133 needed?

LPeterson18 replied:
For whatever reason, it doesn't recognize that location is being used.  If I leave it in I can get the page url, but it triggers a linting error.  When I remove it I am no longer able to get the page url, but the error goes away.

algtan reviewed (changes requested):
Please see comments for requested changes/items for clarification.

LPeterson18 commented:
Moved uri declaration into try catch.

LPeterson18 commented:
Removed unnecessary variable declaration
